source/agora/common/Deserializer.d

```diff
@@ -188,35 +193,113 @@ public void deserializePart (ref ubyte record, scope DeserializeDg dg)
 public void deserializePart (ref ushort record, scope DeserializeDg dg)
     @trusted
 {
-    record = *cast(ushort*)(dg(record.sizeof).ptr);
+    deserializeVarInt(record, dg);
 }
 
 /// Ditto
 public void deserializePart (ref uint record, scope DeserializeDg dg)
     @trusted
 {
-    record = *cast(uint*)(dg(record.sizeof).ptr);
+    deserializeVarInt(record, dg);
 }
 
 /// Ditto
-public void deserializePart (ref ulong record, scope DeserializeDg dg)
+public void deserializePart (ref ulong record, scope DeserializeDg dg, CompactMode compact = CompactMode.Yes)
     @trusted
 {
-    record = *cast(ulong*)(dg(record.sizeof).ptr);
+    if(compact)
+        deserializeVarInt(record, dg);
+    else
+        record = *cast(ulong*)(dg(record.sizeof).ptr);
 }
 
 /// Ditto
 public void deserializePart (ref char[] record, scope DeserializeDg dg)
     @trusted
 {
-    auto length = *cast(size_t*)(dg(size_t.sizeof).ptr);
+    size_t length;
+    deserializeVarInt(length, dg);
     record = cast(char[])dg(length);
 }
 
 /// Ditto
 public void deserializePart (ref ubyte[] record, scope DeserializeDg dg)
     @trusted
 {
-    auto length = *cast(size_t*)(dg(size_t.sizeof).ptr);
+    size_t length;
+    deserializeVarInt(length, dg);
     record = dg(length);
 }
+
+/*******************************************************************************
+
+    Deserialize varint a struct and convert T type return it.
+
+    VarInt Size
+    size_tag is first a ubyte
+    size_tag <= 253(0xFC)  -- 1 byte   ubyte
+    size_tag == 0xFD       -- 3 bytes  (0xFD + ushort)
+    size_tag == 0xFE       -- 5 bytes  (0xFE + uint)
+    size_tag == 0xFF       -- 9 bytes  (0xFF + ulong)
+
+    Params:
+        T = Type of struct to deserialize
+        record = Binary serialized representation of `T`
```

Comment:
We should rename the parameter. It's not a record (a struct), it's an integer. Maybe just call it `value` instead, or `num` for number.

Reply:
use `num`name.